api/osb/reference_instance_plugin.go

```diff
@@ -0,0 +1,361 @@
+
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	return p.generateOSBResponse(ctx, "Deprovision", nil)
+}
+
+// UpdateService intercepts update service instance requests and check if the instance is in the platform from where the request comes
+func (p *referenceInstancePlugin) UpdateService(req *web.Request, next web.Handler) (*web.Response, error) {
+	// we don't want to allow plan_id and/or parameters changes on a reference service instance
+	resourceID := req.PathParams["resource_id"]
+	if resourceID == "" {
+		return next.Handle(req)
+	}
+	ctx := req.Context()
+
+	dbInstanceObject, err := p.getObjectByOperator(ctx, types.ServiceInstanceType, "id", resourceID)
+	if err != nil {
+		return nil, util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+
+	isReferencePlan, err := p.isReferencePlan(ctx, planIDProperty, instance.ServicePlanID)
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	_, err = p.isValidPatchRequest(req, instance)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.generateOSBResponse(ctx, "UpdateService", nil)
+}
+
+// Bind intercepts bind requests and check if the instance is in the platform from where the request comes
+func (p *referenceInstancePlugin) Bind(req *web.Request, next web.Handler) (*web.Response, error) {
+	return p.handleBinding(req, next)
+}
+
+// Unbind intercepts unbind requests and check if the instance is in the platform from where the request comes
+func (p *referenceInstancePlugin) Unbind(req *web.Request, next web.Handler) (*web.Response, error) {
+	return p.handleBinding(req, next)
+}
+
+// FetchBinding intercepts get service binding requests and check if the instance owner is the same as the one requesting the operation
+func (p *referenceInstancePlugin) FetchBinding(req *web.Request, next web.Handler) (*web.Response, error) {
+	return p.handleBinding(req, next)
+}
+
+func (p *referenceInstancePlugin) FetchService(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	instanceID := req.PathParams["instance_id"]
+
+	dbInstanceObject, err := p.getObjectByOperator(ctx, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+
+	isReferencePlan, err := p.isReferencePlan(ctx, "id", instance.ServicePlanID)
+
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	return p.generateOSBResponse(ctx, "FetchService", instance)
+}
+
+func (p *referenceInstancePlugin) generateOSBResponse(ctx context.Context, method string, instance *types.ServiceInstance) (*web.Response, error) {
+	var marshal []byte
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	switch method {
+	case "Provision":
```

Comment:
do we have a test case for each method ?

Reply:
We have the test "fetches reference instance successfully", which also checks:
`referencePlan := GetReferencePlanOfExistingPlan(ctx, "catalog_id", shareablePlanCatalogID)
resp.JSON().Object().Value("plan_id").String().Contains(referencePlan.ID)
resp.JSON().Object().Value("service_id").String().Contains(service2CatalogID)			resp.JSON().Object().Value("parameters").Object().Value("referenced_instance_id").String().Contains(sharedInstanceID)`
-- referenceInstanceID - just added now
Let me know if you think that more validations here

I will add another test to cover the default.